---
title: "BeviMed Introduction"
author: "Daniel Greene"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{BeviMed Introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---
```{r echo=FALSE}
knitr::opts_chunk$set(dev="svg", fig.width=7, fig.height=7)
```

`BeviMed`, which stands for *Bayesian Evaluation of Variant Involvement in Mendelian Disease*, is an association test which estimates both the probability of an association between a given set of variants and a case/control disease label, and, given the association, the probability that each individual variant is pathogenic with respect to the disease. 

The inference is carried out based on the inputs:

* case-control label `y`, a length `N` (number of samples) `logical` vector, 
* genotype matrix `G`, an `N` by `k` integer matrix of allele counts for `N` individuals at `k` rare variant sites, 
* a minimum allele count `min_ac`, representing a mode of inheritance hypothesis (i.e. minimum number of pathogenic variants required to be considered to have a *pathogenic configuration* of variants).

Then, depending on the quantity of interest, the inference procedure can be invoked simply by passing the above arguments to the functions:

* `prob_association` - returning the probability of association between configurations of variants represented in `G` and the case-control label `y` (optionally broken down by mode of inheritance). By default, the prior probability of association is `0.01`, and the prior probabilities of dominant and recessive inheritance given that there is an association are each `0.5`.
* `log_BF` - the log Bayes factor between the association model and no-association model.
* `prob_pathogenic` - the probabilities of pathogenicity for the individual variants. 

The inference is performed by the function `bevimed`, an MCMC sampling procedure with many parameters, including those listed above and others determining the sampling management and prior distributions of the model parameters. 

It returns a list of traces for the sampled parameters in an object of class `BeviMed`. This object can take up a lot of memory, so it may be preferable to store a summarised version passed to `summary`. 

Here we demonstrate a simple application of BeviMed for some simulated data.

```{r}
library(BeviMed)
set.seed(0)
```

Firstly, we'll generate some random data consisting of an allele-count matrix `G` for 100 samples at 20 variant sites (each with an allele frequency of 0.02) and an independently generated case-control label, `y_random`.

```{r}
G <- matrix(rbinom(size=2, prob=0.02, n=100*20), nrow=100, ncol=20)
y_random <- runif(n=nrow(G)) < 0.1

prob_association(G=G, y=y_random)
```

The results indicate that there is a low probability of association. We now generate a new case control label `y_dependent` which depends on `G` - specifically, we treat variants 1 to 3 as 'pathogenic', and label any samples harbouring alleles for any of these variants as cases.

```{r}
y_dependent <- apply(G, 1, function(variants) sum(variants[1:3]) > 0)

prob_association(G=G, y=y_dependent)
```

Notice that there is now a higher estimated probability of association. 

By default, `prob_association` integrates over mode of inheritance (e.g. are at least 1 or 2 pathogenic variants required for a pathogenic configuration?). The probabilities of association with each mode of inheritance can by shown by passing the option `by_MOI=TRUE` (for more details, including how to set the ploidy of the samples within the region, see `?prob_pathogenic`).

For a more detailed output, the `bevimed` function can be used, and it's returned values can be summarised and stored/printed.

```{r}
output <- summary(bevimed(G=G, y=y_dependent))
output
```
