<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Daniel Greene" />

<meta name="date" content="2025-04-29" />

<title>BeviMed Introduction</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">BeviMed Introduction</h1>
<h4 class="author">Daniel Greene</h4>
<h4 class="date">2025-04-29</h4>



<p><code>BeviMed</code>, which stands for <em>Bayesian Evaluation of
Variant Involvement in Mendelian Disease</em> [1], is an association
test which estimates the probability of an association between a given
set of variants and a case/control label, the mode of inheritance for
the disease, and the probability that each individual variant is
pathogenic. This vignette gives a quick description of what can be done
with the package and demonstrates how to use it on simulated data. For
more detailed explanations, see the ‘BeviMed Guide’ vignette and the
individual function help pages.</p>
<p>Inference is performed by the <code>bevimed</code> function, which
evaluates the data with respect to three models: a model of no
association between case/control label and allele counts, and models of
dominant and recessive association. The function depends on input
parameters:</p>
<ul>
<li>case-control label <code>y</code>, a length <code>N</code> (number
of samples) <code>logical</code> vector,</li>
<li>genotype matrix <code>G</code>, an <code>N</code> by <code>k</code>
integer matrix of allele counts for <code>N</code> individuals at
<code>k</code> rare variant sites,</li>
<li><code>prior_prob_association</code> - the prior probability of
association between the disease label and the variants in the locus.
Defaults to <code>0.01</code>.</li>
<li><code>prior_prob_dominant</code> - the prior probability of dominant
as opposed to recessive inheritance, given an association with the
locus. Defaults to <code>0.5</code>.</li>
<li><code>ploidy</code> - a vector the same length as the case-control
label <code>y</code> giving the ploidy of each individual in the locus.
Defaults to <code>2</code> for each sample.</li>
<li>other arguments controlling the inference procedure and prior
distributions of parameters (see <code>?bevimed</code> for more
details).</li>
</ul>
<p><code>bevimed</code> returns an object of class <code>BeviMed</code>
which contains the output of the (MCMC-based) inference procedure,
including samples from the posterior distributions of the model
parameters. The object can be evaluated at the command line to print a
summary of inference, telling you summary statistics of interest,
including the probability of association. The object is likely to take
up a lot of memory, so it is useful to store a summary, computed with
<code>summary</code>, for each result if the procedure is being applied
to multiple loci.</p>
<p>Summary statistics can also be computed directly from the arguments
using the functions (see help for individual functions for more
information):</p>
<ul>
<li><code>prob_association</code> - returning the probability of
association, optionally broken down by mode of inheritance.</li>
<li><code>conditional_prob_pathogenic</code> - the probabilities of
pathogenicity for the individual variants conditional on a mode of
inheritance.</li>
<li><code>expected_explained</code> - the expected number of cases
explained by variants.</li>
<li><code>explaining_variants</code> - the expected number of variants
involved in explained cases.</li>
</ul>
<p>Here we demonstrate a simple application of BeviMed for some
simulated data.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(BeviMed)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">0</span>)</span></code></pre></div>
<p>Firstly, we’ll generate a random allele-count matrix <code>G</code>
for 100 samples at 20 variant sites (each with an allele frequency of
0.02) and an independently generated case-control label,
<code>y_random</code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>G <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rbinom</span>(<span class="at">size=</span><span class="dv">2</span>, <span class="at">prob=</span><span class="fl">0.02</span>, <span class="at">n=</span><span class="dv">100</span><span class="sc">*</span><span class="dv">20</span>), <span class="at">nrow=</span><span class="dv">100</span>, <span class="at">ncol=</span><span class="dv">20</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>y_random <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="at">n=</span><span class="fu">nrow</span>(G)) <span class="sc">&lt;</span> <span class="fl">0.1</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="fu">prob_association</span>(<span class="at">G=</span>G, <span class="at">y=</span>y_random)</span></code></pre></div>
<pre><code>## [1] 0.002810699</code></pre>
<p>The results indicate that there is a low probability of association.
We now generate a new case control label <code>y_dependent</code> which
depends on <code>G</code> - specifically, we treat variants 1 to 3 as
‘pathogenic’, and label any samples harbouring alleles for any of these
variants as cases.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>y_dependent <span class="ot">&lt;-</span> <span class="fu">apply</span>(G, <span class="dv">1</span>, <span class="cf">function</span>(variants) <span class="fu">sum</span>(variants[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]) <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">prob_association</span>(<span class="at">G=</span>G, <span class="at">y=</span>y_dependent)</span></code></pre></div>
<pre><code>## [1] 0.9998203</code></pre>
<p>Notice that there is now a higher estimated probability of
association.</p>
<p>By default, <code>prob_association</code> integrates over mode of
inheritance (e.g. are at least 1 or 2 pathogenic variants required for a
pathogenic configuration?). The probabilities of association with each
mode of inheritance can by shown by passing the option
<code>by_MOI=TRUE</code> (for more details, including how to set the
ploidy of the samples within the region, see
<code>?prob_pathogenic</code>).</p>
<p>For a more detailed output, the <code>bevimed</code> function can be
used, and it’s returned values can be summarised and stored/printed.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">summary</span>(<span class="fu">bevimed</span>(<span class="at">G=</span>G, <span class="at">y=</span>y_dependent))</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>output</span></code></pre></div>
<pre><code>## -------------------------------------------------------------------------------- 
## Posterior probability of association: 
##  1 [prior: 0.01]
## -------------------------------------------------------------------------------- 
##      Model MOI Prior     Post Cases Variants
##   dominant dom   0.5 0.999676  6.93     2.95
##  recessive rec   0.5 0.000324  4.59     5.96
## 
## MOI: mode of inheritance, dominant (dom) or recessive (rec)
## Prior: prior probability of model given association
## Post: posterior probability of model given association
## Cases: posterior expected number of cases explained
## Variants: posterior expected number of variants involved in explained cases
## -------------------------------------------------------------------------------- 
## Probabilities of pathogenicity for individual variants given association
## 
##  Var   Probability pathogenic
##    1 [1.00 =================]
##    2 [1.00 =================]
##    3 [0.94 ================ ]
##    4 [0.00                  ]
##    5 [0.00                  ]
##    6 [0.00                  ]
##    7 [0.00                  ]
##    8 [0.00                  ]
##    9 [0.00                  ]
##   10 [0.00                  ]
##   11 [0.00                  ]
##   12 [0.01                  ]
##   13 [0.00                  ]
##   14 [0.02                  ]
##   15 [0.00                  ]
##   16 [0.00                  ]
##   17 [0.00                  ]
##   18 [0.00                  ]
##   19 [0.00                  ]
##   20 [0.00                  ]
## --------------------------------------------------------------------------------</code></pre>
<div id="references" class="section level2">
<h2>References</h2>
<ol style="list-style-type: decimal">
<li>Greene et al., A Fast Association Test for Identifying Pathogenic
Variants Involved in Rare Diseases, The American Journal of Human
Genetics (2017), <a href="http://dx.doi.org/10.1016/j.ajhg.2017.05.015" class="uri">http://dx.doi.org/10.1016/j.ajhg.2017.05.015</a></li>
</ol>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
