<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Daniel Greene" />

<meta name="date" content="2016-05-02" />

<title>BeviMed Introduction</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title">BeviMed Introduction</h1>
<h4 class="author"><em>Daniel Greene</em></h4>
<h4 class="date"><em>2016-05-02</em></h4>
</div>


<p><code>BeviMed</code> has a function <code>bevimed</code> for performing <em>Bayesian Evaluation of Variant Involvement in Mendelian Disease</em>. The aim is to estimate the probability of an association between individual configurations of a restricted set of variants and a case-control label, and conditional on the association, the probabilities of pathogenicity for individual variants.</p>
<p><code>bevimed</code> is an MCMC sampling procedure with many parameters, including the data <code>y</code> (a <code>logical</code> vector), a <code>k</code> by <code>N</code> matrix of variant-level genotypes <code>G</code> (where <code>length(y) == nrow(G)</code>), a mode of inheritance hypothesis encoded as an integer (minimum number of alleles required for affection) <code>min_ac</code>, and others determining the sampling management and prior distributions of the model parameters.</p>
<p>It returns a list of traces for the sampled parameters in an object of class <code>BeviMed</code>. This object can take up a lot of, and hence it may be preferable to store a summarised version by passing it to <code>summary</code>.</p>
<p>Quantities of interest can also be computed directly from the parameters - for example via the function <code>log BF</code>, returning a log Bayes factor between the variant-level model for case-control status and a baseline model whereby there is no association between variants in the region and case-control status - or the function <code>probability_pathogenic</code> a probability of association given some prior probability.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(BeviMed)
<span class="kw">set.seed</span>(<span class="dv">0</span>)</code></pre>
<p>Firstly, weâ€™ll generate some random data consisting of a disease affection status, <code>y</code> and a variant-wise indicator of pathogenicity, <code>Z</code>, where <code>Z[i] == TRUE</code> if variant <code>i</code> is pathogenic and <code>FALSE</code> otherwise.</p>
<pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">rep</span>(<span class="ot">TRUE</span>, <span class="dv">20</span>), <span class="kw">rep</span>(<span class="ot">FALSE</span>, <span class="dv">80</span>))
Z &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">rep</span>(<span class="ot">TRUE</span>, <span class="dv">3</span>), <span class="kw">rep</span>(<span class="ot">FALSE</span>, <span class="dv">20</span>))</code></pre>
<p>In the first application, weâ€™ll use variant matrix <code>G1</code> where there is no association between <code>y</code> and the variants.</p>
<pre class="sourceCode r"><code class="sourceCode r">G1 &lt;-<span class="st"> </span><span class="kw">sapply</span>(y, function(y_i) <span class="kw">as.integer</span>(<span class="kw">runif</span>(<span class="dt">n=</span><span class="kw">length</span>(Z)) &lt;<span class="st"> </span><span class="fl">0.15</span>))

<span class="kw">probability_pathogenic</span>(<span class="dt">G=</span>G1, <span class="dt">y=</span>y)</code></pre>
<pre><code>## [1] 0.06988054</code></pre>
<p>The results indicate that there is little preference for model <span class="math">\(v\)</span> over model <span class="math">\(n\)</span>. In the example <em>with</em> association, weâ€™ll sample the allele counts based on <code>y</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">G2 &lt;-<span class="st"> </span><span class="kw">sapply</span>(y, function(y_i) <span class="kw">as.integer</span>(<span class="kw">runif</span>(<span class="dt">n=</span><span class="kw">length</span>(Z)) &lt;<span class="st"> </span><span class="fl">0.15</span> |<span class="st"> </span>
<span class="st">    </span>(if (y_i) <span class="dv">1</span>:<span class="kw">length</span>(Z) ==<span class="st"> </span><span class="kw">sample</span>(<span class="kw">which</span>(Z), <span class="dt">size=</span><span class="dv">1</span>) else <span class="kw">rep</span>(<span class="ot">FALSE</span>, <span class="kw">length</span>(Z)))))

<span class="kw">probability_pathogenic</span>(<span class="dt">G=</span>G2, <span class="dt">y=</span>y)</code></pre>
<pre><code>## [1] 0.9999853</code></pre>
<p>For a more detailed output (containing the variant-level probabilities of pathogenicity), the <code>bevimed</code> function can be called, summarised and stored.</p>
<pre class="sourceCode r"><code class="sourceCode r">output &lt;-<span class="st"> </span><span class="kw">summary</span>(<span class="kw">bevimed</span>(<span class="dt">G=</span>G2, <span class="dt">y=</span>y))

output</code></pre>
<pre><code>## --------------------------------------------------------------------------- 
## Log Bayes factor between variant level model and null model is 13.94
## 
## A confidence interval for the estimate is:
##  2.5% 97.5% 
## 13.26 14.49 
## ---------------------------------------------------------------------------</code></pre>
<p>Note that now the estimated log Bayes factor in favour of the variant-level model is high. Furthermore, we can use <code>print</code> passing the option <code>print_Z=TRUE</code> to print a summary of the results containing the variant-level probabilities of pathogenicity.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(output, <span class="dt">print_Z=</span><span class="ot">TRUE</span>)</code></pre>
<pre><code>## --------------------------------------------------------------------------- 
## Log Bayes factor between variant level model and null model is 13.94
## 
## A confidence interval for the estimate is:
##  2.5% 97.5% 
## 13.26 14.49 
## --------------------------------------------------------------------------- 
## Estimated probabilities of pathogenicity of individual variants
## (conditional on model v)
## 
##  Variant Controls Cases P(Z_j=1|y,V)              Bar Chart
##        1       15     8         0.99 [=================== ]
##        2       10    11         1.00 [====================]
##        3        6     5         0.92 [==================  ]
##        4       12     8         0.00 [                    ]
##        5       15     3         0.00 [                    ]
##        6       11     1         0.00 [                    ]
##        7       14     5         0.00 [                    ]
##        8       15     3         0.00 [                    ]
##        9       15     2         0.00 [                    ]
##       10        8     2         0.01 [                    ]
##       11       15     2         0.00 [                    ]
##       12        8     4         0.01 [                    ]
##       13       13     1         0.00 [                    ]
##       14        9     3         0.04 [                    ]
##       15       10     3         0.00 [                    ]
##       16       16     1         0.00 [                    ]
##       17       14     2         0.00 [                    ]
##       18       13     3         0.00 [                    ]
##       19       12     1         0.00 [                    ]
##       20       17     2         0.00 [                    ]
##       21       13     7         0.00 [                    ]
##       22       10     3         0.01 [                    ]
##       23       15     6         0.00 [                    ]</code></pre>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
